--- ../src-base/minecraft/net/minecraft/entity/item/EntityItemFrame.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityItemFrame.java
@@ -1,6 +1,9 @@
 package net.minecraft.entity.item;
 
 import javax.annotation.Nullable;
+
+import net.minecraft.block.BlockRedstoneDiode;
+import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityHanging;
 import net.minecraft.entity.player.EntityPlayer;
@@ -19,11 +22,13 @@
 import net.minecraft.util.datafix.DataFixer;
 import net.minecraft.util.datafix.FixTypes;
 import net.minecraft.util.datafix.walkers.ItemStackData;
+import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 import net.minecraft.world.storage.MapData;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.apache.commons.lang3.Validate;
 
 public class EntityItemFrame extends EntityHanging
 {
@@ -48,6 +53,100 @@
         this.func_184212_Q().func_187214_a(field_184526_d, Integer.valueOf(0));
     }
 
+    @Override
+    protected void func_174859_a(EnumFacing facingDirectionIn) {
+        if (facingDirectionIn.func_176740_k().func_176722_c()) {
+            super.func_174859_a(facingDirectionIn);
+        } else {
+            //facing up
+            if (facingDirectionIn == EnumFacing.UP) {
+                this.func_70105_a(0F, 2F);
+            }
+            if (facingDirectionIn == EnumFacing.DOWN) {
+                this.func_70105_a(0F, -2F);
+            }
+            Validate.notNull(facingDirectionIn);
+            //Validate.isTrue(facingDirectionIn.getAxis().isHorizontal());
+            this.field_174860_b = facingDirectionIn;
+
+            this.field_70125_A = (float) (90 * facingDirectionIn.func_176743_c().func_179524_a());
+            this.field_70177_z = 90.0F;
+
+            this.field_70127_C = this.field_70125_A;
+            this.field_70126_B = this.field_70177_z;
+            //  side    index   axisoff prPitch     prYaw
+            //  up      1       1       -90         0
+            //  down    0       -1      90          0
+            
+            
+      /*      System.out.println("ishorizontal: " + facingDirectionIn.getAxis().isHorizontal());
+            System.out.println("facing index : " + this.facingDirection.getIndex());
+            System.out.println("facingIn axis offset: " + facingDirectionIn.getAxisDirection().getOffset());
+            System.out.println("prevRotationPitch: " + this.prevRotationPitch);
+            System.out.println("prevRotationYaw: " + this.prevRotationYaw);*/
+
+            this.func_174856_o();
+         
+        }
+    }
+    
+    @Override
+    protected void func_174856_o() {
+
+        
+        if (this.field_174860_b != null) {
+            if (this.field_174860_b.func_176740_k().func_176722_c()) {
+                super.func_174856_o();
+            } else {
+                double x = (double) this.field_174861_a.func_177958_n() + 0.5D - (double) this.field_174860_b.func_82601_c() * 0.46875D;
+                double y = (double) this.field_174861_a.func_177956_o() + 0.5D - (double) this.field_174860_b.func_96559_d() * 0.46875D;
+                double z = (double) this.field_174861_a.func_177952_p() + 0.5D - (double) this.field_174860_b.func_82599_e() * 0.46875D;
+                this.field_70165_t = x;
+                this.field_70163_u = y;
+                this.field_70161_v = z;
+                double x1 = (double) this.func_82329_d();
+                double y1 = (double) this.func_82330_g();
+                double z1 = (double) this.func_82329_d();
+
+                switch (this.field_174860_b.func_176740_k()) {
+                    case X:
+                        x1 = 1.0D;
+                        break;
+                    case Y:
+                        y1 = 1.0D;
+                        break;
+                    case Z:
+                        z1 = 1.0D;
+                        break;
+                }
+                x1 /= 32.0D;
+                y1 /= 32.0D;
+                z1 /= 32.0D;
+                //System.out.println("getAxis" + this.facingDirection.getAxis());
+                //System.out.println("x,y,z,x1,y1,z1: "+x+","+y+","+z+","+x1+","+y1+","+z1);
+                this.func_174826_a(new AxisAlignedBB(x - x1, y - y1, z - z1, x + x1, y + y1, z + z1));
+            }
+          
+
+        }
+    }
+    
+    @Override
+    public boolean func_70518_d() {
+
+        if (!this.field_70170_p.func_184144_a(this, this.func_174813_aQ()).isEmpty())
+        {
+            return false;
+        }
+        else
+        {
+            IBlockState blockState = this.field_70170_p.func_180495_p(this.field_174861_a.func_177972_a(this.field_174860_b.func_176734_d()));
+            return blockState.func_185904_a().func_76220_a() || this.field_174860_b.func_176740_k().func_176722_c() &&
+                    BlockRedstoneDiode.func_185546_B(blockState) ? this.field_70170_p.func_175674_a(this, this.func_174813_aQ(), field_184524_c).isEmpty() : false;
+            
+        }
+    }
+
     public float func_70111_Y()
     {
         return 0.0F;
@@ -140,13 +239,14 @@
     {
         if (!p_110131_1_.func_190926_b())
         {
-            if (p_110131_1_.func_77973_b() == Items.field_151098_aY)
+            if (p_110131_1_.func_77973_b() instanceof net.minecraft.item.ItemMap)
             {
                 MapData mapdata = ((ItemMap)p_110131_1_.func_77973_b()).func_77873_a(p_110131_1_, this.field_70170_p);
                 mapdata.field_76203_h.remove("frame-" + this.func_145782_y());
             }
 
             p_110131_1_.func_82842_a((EntityItemFrame)null);
+            this.func_82334_a(ItemStack.field_190927_a); //Forge: Fix MC-124833 Pistons duplicating Items.
         }
     }
 
@@ -223,18 +323,25 @@
 
     public void func_70014_b(NBTTagCompound p_70014_1_)
     {
+        //cherry update
+        super.func_70014_b(p_70014_1_);
         if (!this.func_82335_i().func_190926_b())
         {
             p_70014_1_.func_74782_a("Item", this.func_82335_i().func_77955_b(new NBTTagCompound()));
             p_70014_1_.func_74774_a("ItemRotation", (byte)this.func_82333_j());
             p_70014_1_.func_74776_a("ItemDropChance", this.field_82337_e);
         }
+        //cherry add save facing
+        p_70014_1_.func_74774_a("Facing", (byte)this.field_174860_b.func_176745_a());
 
-        super.func_70014_b(p_70014_1_);
     }
 
     public void func_70037_a(NBTTagCompound p_70037_1_)
     {
+
+        //cherry modify
+        super.func_70037_a(p_70037_1_);
+        
         NBTTagCompound nbttagcompound = p_70037_1_.func_74775_l("Item");
 
         if (nbttagcompound != null && !nbttagcompound.func_82582_d())
@@ -247,8 +354,10 @@
                 this.field_82337_e = p_70037_1_.func_74760_g("ItemDropChance");
             }
         }
+        
+        //cherry add   read facing
+        this.func_174859_a(EnumFacing.func_82600_a(p_70037_1_.func_74771_c("Facing")));
 
-        super.func_70037_a(p_70037_1_);
     }
 
     public boolean func_184230_a(EntityPlayer p_184230_1_, EnumHand p_184230_2_)
