--- ../src-base/minecraft/net/minecraft/client/gui/GuiOverlayDebug.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiOverlayDebug.java
@@ -189,6 +189,9 @@
         long l = j - k;
         List<String> list = Lists.newArrayList(String.format("Java: %s %dbit", System.getProperty("java.version"), this.field_175242_a.func_147111_S() ? 64 : 32), String.format("Mem: % 2d%% %03d/%03dMB", l * 100L / i, func_175240_a(l), func_175240_a(i)), String.format("Allocated: % 2d%% %03dMB", j * 100L / i, func_175240_a(j)), "", String.format("CPU: %s", OpenGlHelper.func_183029_j()), "", String.format("Display: %dx%d (%s)", Display.getWidth(), Display.getHeight(), GlStateManager.func_187416_u(7936)), GlStateManager.func_187416_u(7937), GlStateManager.func_187416_u(7938));
 
+        list.add("");
+        list.addAll(net.minecraftforge.fml.common.FMLCommonHandler.instance().getBrandings(false));
+
         if (this.field_175242_a.func_189648_am())
         {
             return list;
@@ -232,44 +235,46 @@
         }
     }
 
-    public void func_181554_e()
+    public void func_181554_e() {
+    }
+   /* public void renderLagometer_china()
     {
-        GlStateManager.func_179097_i();
-        FrameTimer frametimer = this.field_175242_a.func_181539_aj();
-        int i = frametimer.func_181749_a();
-        int j = frametimer.func_181750_b();
-        long[] along = frametimer.func_181746_c();
-        ScaledResolution scaledresolution = new ScaledResolution(this.field_175242_a);
+        GlStateManager.disableDepth();
+        FrameTimer frametimer = this.mc.getFrameTimer();
+        int i = frametimer.getLastIndex();
+        int j = frametimer.getIndex();
+        long[] along = frametimer.getFrames();
+        ScaledResolution scaledresolution = new ScaledResolution(this.mc);
         int k = i;
         int l = 0;
-        func_73734_a(0, scaledresolution.func_78328_b() - 60, 240, scaledresolution.func_78328_b(), -1873784752);
+        drawRect(0, scaledresolution.getScaledHeight() - 60, 240, scaledresolution.getScaledHeight(), -1873784752);
 
         while (k != j)
         {
-            int i1 = frametimer.func_181748_a(along[k], 30);
-            int j1 = this.func_181552_c(MathHelper.func_76125_a(i1, 0, 60), 0, 30, 60);
-            this.func_73728_b(l, scaledresolution.func_78328_b(), scaledresolution.func_78328_b() - i1, j1);
+            int i1 = frametimer.getLagometerValue(along[k], 30);
+            int j1 = this.getFrameColor(MathHelper.clamp(i1, 0, 60), 0, 30, 60);
+            this.drawVerticalLine(l, scaledresolution.getScaledHeight(), scaledresolution.getScaledHeight() - i1, j1);
             ++l;
-            k = frametimer.func_181751_b(k + 1);
+            k = frametimer.parseIndex(k + 1);
         }
 
-        func_73734_a(1, scaledresolution.func_78328_b() - 30 + 1, 14, scaledresolution.func_78328_b() - 30 + 10, -1873784752);
-        this.field_175241_f.func_78276_b("60", 2, scaledresolution.func_78328_b() - 30 + 2, 14737632);
-        this.func_73730_a(0, 239, scaledresolution.func_78328_b() - 30, -1);
-        func_73734_a(1, scaledresolution.func_78328_b() - 60 + 1, 14, scaledresolution.func_78328_b() - 60 + 10, -1873784752);
-        this.field_175241_f.func_78276_b("30", 2, scaledresolution.func_78328_b() - 60 + 2, 14737632);
-        this.func_73730_a(0, 239, scaledresolution.func_78328_b() - 60, -1);
-        this.func_73730_a(0, 239, scaledresolution.func_78328_b() - 1, -1);
-        this.func_73728_b(0, scaledresolution.func_78328_b() - 60, scaledresolution.func_78328_b(), -1);
-        this.func_73728_b(239, scaledresolution.func_78328_b() - 60, scaledresolution.func_78328_b(), -1);
+        drawRect(1, scaledresolution.getScaledHeight() - 30 + 1, 14, scaledresolution.getScaledHeight() - 30 + 10, -1873784752);
+        this.fontRenderer.drawString("60", 2, scaledresolution.getScaledHeight() - 30 + 2, 14737632);
+        this.drawHorizontalLine(0, 239, scaledresolution.getScaledHeight() - 30, -1);
+        drawRect(1, scaledresolution.getScaledHeight() - 60 + 1, 14, scaledresolution.getScaledHeight() - 60 + 10, -1873784752);
+        this.fontRenderer.drawString("30", 2, scaledresolution.getScaledHeight() - 60 + 2, 14737632);
+        this.drawHorizontalLine(0, 239, scaledresolution.getScaledHeight() - 60, -1);
+        this.drawHorizontalLine(0, 239, scaledresolution.getScaledHeight() - 1, -1);
+        this.drawVerticalLine(0, scaledresolution.getScaledHeight() - 60, scaledresolution.getScaledHeight(), -1);
+        this.drawVerticalLine(239, scaledresolution.getScaledHeight() - 60, scaledresolution.getScaledHeight(), -1);
 
-        if (this.field_175242_a.field_71474_y.field_74350_i <= 120)
+        if (this.mc.gameSettings.limitFramerate <= 120)
         {
-            this.func_73730_a(0, 239, scaledresolution.func_78328_b() - 60 + this.field_175242_a.field_71474_y.field_74350_i / 2, -16711681);
+            this.drawHorizontalLine(0, 239, scaledresolution.getScaledHeight() - 60 + this.mc.gameSettings.limitFramerate / 2, -16711681);
         }
 
-        GlStateManager.func_179126_j();
-    }
+        GlStateManager.enableDepth();
+    }*/
 
     private int func_181552_c(int p_181552_1_, int p_181552_2_, int p_181552_3_, int p_181552_4_)
     {
