--- ../src-base/minecraft/net/minecraft/client/renderer/BlockFluidRenderer.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/BlockFluidRenderer.java
@@ -2,6 +2,7 @@
 
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockLiquid;
+import net.minecraft.block.BlockSlab;
 import net.minecraft.block.material.Material;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.client.Minecraft;
@@ -15,6 +16,10 @@
 import net.minecraft.world.IBlockAccess;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import optifine.Config;
+import optifine.CustomColors;
+import optifine.RenderEnv;
+import shadersmod.client.SVertexBuilder;
 
 @SideOnly(Side.CLIENT)
 public class BlockFluidRenderer
@@ -40,18 +45,18 @@
         this.field_187501_d = texturemap.func_110572_b("minecraft:blocks/water_overlay");
     }
 
-    public boolean func_178270_a(IBlockAccess p_178270_1_, IBlockState p_178270_2_, BlockPos p_178270_3_, BufferBuilder p_178270_4_)
+    public boolean renderFluidO(IBlockAccess blockAccess, IBlockState blockStateIn, BlockPos blockPosIn, BufferBuilder bufferBuilderIn)
     {
-        BlockLiquid blockliquid = (BlockLiquid)p_178270_2_.func_177230_c();
-        boolean flag = p_178270_2_.func_185904_a() == Material.field_151587_i;
+        BlockLiquid blockliquid = (BlockLiquid)blockStateIn.func_177230_c();
+        boolean flag = blockStateIn.func_185904_a() == Material.field_151587_i;
         TextureAtlasSprite[] atextureatlassprite = flag ? this.field_178272_a : this.field_178271_b;
-        int i = this.field_187500_a.func_186724_a(p_178270_2_, p_178270_1_, p_178270_3_, 0);
+        int i = this.field_187500_a.func_186724_a(blockStateIn, blockAccess, blockPosIn, 0);
         float f = (float)(i >> 16 & 255) / 255.0F;
         float f1 = (float)(i >> 8 & 255) / 255.0F;
         float f2 = (float)(i & 255) / 255.0F;
-        boolean flag1 = p_178270_2_.func_185894_c(p_178270_1_, p_178270_3_, EnumFacing.UP);
-        boolean flag2 = p_178270_2_.func_185894_c(p_178270_1_, p_178270_3_, EnumFacing.DOWN);
-        boolean[] aboolean = new boolean[] {p_178270_2_.func_185894_c(p_178270_1_, p_178270_3_, EnumFacing.NORTH), p_178270_2_.func_185894_c(p_178270_1_, p_178270_3_, EnumFacing.SOUTH), p_178270_2_.func_185894_c(p_178270_1_, p_178270_3_, EnumFacing.WEST), p_178270_2_.func_185894_c(p_178270_1_, p_178270_3_, EnumFacing.EAST)};
+        boolean flag1 = blockStateIn.func_185894_c(blockAccess, blockPosIn, EnumFacing.UP);
+        boolean flag2 = blockStateIn.func_185894_c(blockAccess, blockPosIn, EnumFacing.DOWN);
+        boolean[] aboolean = new boolean[] {blockStateIn.func_185894_c(blockAccess, blockPosIn, EnumFacing.NORTH), blockStateIn.func_185894_c(blockAccess, blockPosIn, EnumFacing.SOUTH), blockStateIn.func_185894_c(blockAccess, blockPosIn, EnumFacing.WEST), blockStateIn.func_185894_c(blockAccess, blockPosIn, EnumFacing.EAST)};
 
         if (!flag1 && !flag2 && !aboolean[0] && !aboolean[1] && !aboolean[2] && !aboolean[3])
         {
@@ -64,20 +69,20 @@
             float f4 = 1.0F;
             float f5 = 0.8F;
             float f6 = 0.6F;
-            Material material = p_178270_2_.func_185904_a();
-            float f7 = this.func_178269_a(p_178270_1_, p_178270_3_, material);
-            float f8 = this.func_178269_a(p_178270_1_, p_178270_3_.func_177968_d(), material);
-            float f9 = this.func_178269_a(p_178270_1_, p_178270_3_.func_177974_f().func_177968_d(), material);
-            float f10 = this.func_178269_a(p_178270_1_, p_178270_3_.func_177974_f(), material);
-            double d0 = (double)p_178270_3_.func_177958_n();
-            double d1 = (double)p_178270_3_.func_177956_o();
-            double d2 = (double)p_178270_3_.func_177952_p();
+            Material material = blockStateIn.func_185904_a();
+            float f7 = this.func_178269_a(blockAccess, blockPosIn, material);
+            float f8 = this.func_178269_a(blockAccess, blockPosIn.func_177968_d(), material);
+            float f9 = this.func_178269_a(blockAccess, blockPosIn.func_177974_f().func_177968_d(), material);
+            float f10 = this.func_178269_a(blockAccess, blockPosIn.func_177974_f(), material);
+            double d0 = (double)blockPosIn.func_177958_n();
+            double d1 = (double)blockPosIn.func_177956_o();
+            double d2 = (double)blockPosIn.func_177952_p();
             float f11 = 0.001F;
 
             if (flag1)
             {
                 flag3 = true;
-                float f12 = BlockLiquid.func_189544_a(p_178270_1_, p_178270_3_, material, p_178270_2_);
+                float f12 = BlockLiquid.func_189544_a(blockAccess, blockPosIn, material, blockStateIn);
                 TextureAtlasSprite textureatlassprite = f12 > -999.0F ? atextureatlassprite[1] : atextureatlassprite[0];
                 f7 -= 0.001F;
                 f8 -= 0.001F;
@@ -118,23 +123,23 @@
                     f20 = textureatlassprite.func_94207_b((double)(8.0F + (-f22 - f21) * 16.0F));
                 }
 
-                int k2 = p_178270_2_.func_185889_a(p_178270_1_, p_178270_3_);
+                int k2 = blockStateIn.func_185889_a(blockAccess, blockPosIn);
                 int l2 = k2 >> 16 & 65535;
                 int i3 = k2 & 65535;
                 float f24 = 1.0F * f;
                 float f25 = 1.0F * f1;
                 float f26 = 1.0F * f2;
-                p_178270_4_.func_181662_b(d0 + 0.0D, d1 + (double)f7, d2 + 0.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f13, (double)f17).func_187314_a(l2, i3).func_181675_d();
-                p_178270_4_.func_181662_b(d0 + 0.0D, d1 + (double)f8, d2 + 1.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f14, (double)f18).func_187314_a(l2, i3).func_181675_d();
-                p_178270_4_.func_181662_b(d0 + 1.0D, d1 + (double)f9, d2 + 1.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f15, (double)f19).func_187314_a(l2, i3).func_181675_d();
-                p_178270_4_.func_181662_b(d0 + 1.0D, d1 + (double)f10, d2 + 0.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f16, (double)f20).func_187314_a(l2, i3).func_181675_d();
+                bufferBuilderIn.func_181662_b(d0 + 0.0D, d1 + (double)f7, d2 + 0.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f13, (double)f17).func_187314_a(l2, i3).func_181675_d();
+                bufferBuilderIn.func_181662_b(d0 + 0.0D, d1 + (double)f8, d2 + 1.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f14, (double)f18).func_187314_a(l2, i3).func_181675_d();
+                bufferBuilderIn.func_181662_b(d0 + 1.0D, d1 + (double)f9, d2 + 1.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f15, (double)f19).func_187314_a(l2, i3).func_181675_d();
+                bufferBuilderIn.func_181662_b(d0 + 1.0D, d1 + (double)f10, d2 + 0.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f16, (double)f20).func_187314_a(l2, i3).func_181675_d();
 
-                if (blockliquid.func_176364_g(p_178270_1_, p_178270_3_.func_177984_a()))
+                if (blockliquid.func_176364_g(blockAccess, blockPosIn.func_177984_a()))
                 {
-                    p_178270_4_.func_181662_b(d0 + 0.0D, d1 + (double)f7, d2 + 0.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f13, (double)f17).func_187314_a(l2, i3).func_181675_d();
-                    p_178270_4_.func_181662_b(d0 + 1.0D, d1 + (double)f10, d2 + 0.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f16, (double)f20).func_187314_a(l2, i3).func_181675_d();
-                    p_178270_4_.func_181662_b(d0 + 1.0D, d1 + (double)f9, d2 + 1.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f15, (double)f19).func_187314_a(l2, i3).func_181675_d();
-                    p_178270_4_.func_181662_b(d0 + 0.0D, d1 + (double)f8, d2 + 1.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f14, (double)f18).func_187314_a(l2, i3).func_181675_d();
+                    bufferBuilderIn.func_181662_b(d0 + 0.0D, d1 + (double)f7, d2 + 0.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f13, (double)f17).func_187314_a(l2, i3).func_181675_d();
+                    bufferBuilderIn.func_181662_b(d0 + 1.0D, d1 + (double)f10, d2 + 0.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f16, (double)f20).func_187314_a(l2, i3).func_181675_d();
+                    bufferBuilderIn.func_181662_b(d0 + 1.0D, d1 + (double)f9, d2 + 1.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f15, (double)f19).func_187314_a(l2, i3).func_181675_d();
+                    bufferBuilderIn.func_181662_b(d0 + 0.0D, d1 + (double)f8, d2 + 1.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f14, (double)f18).func_187314_a(l2, i3).func_181675_d();
                 }
             }
 
@@ -144,13 +149,13 @@
                 float f36 = atextureatlassprite[0].func_94212_f();
                 float f37 = atextureatlassprite[0].func_94206_g();
                 float f38 = atextureatlassprite[0].func_94210_h();
-                int l1 = p_178270_2_.func_185889_a(p_178270_1_, p_178270_3_.func_177977_b());
+                int l1 = blockStateIn.func_185889_a(blockAccess, blockPosIn.func_177977_b());
                 int i2 = l1 >> 16 & 65535;
                 int j2 = l1 & 65535;
-                p_178270_4_.func_181662_b(d0, d1, d2 + 1.0D).func_181666_a(0.5F, 0.5F, 0.5F, 1.0F).func_187315_a((double)f35, (double)f38).func_187314_a(i2, j2).func_181675_d();
-                p_178270_4_.func_181662_b(d0, d1, d2).func_181666_a(0.5F, 0.5F, 0.5F, 1.0F).func_187315_a((double)f35, (double)f37).func_187314_a(i2, j2).func_181675_d();
-                p_178270_4_.func_181662_b(d0 + 1.0D, d1, d2).func_181666_a(0.5F, 0.5F, 0.5F, 1.0F).func_187315_a((double)f36, (double)f37).func_187314_a(i2, j2).func_181675_d();
-                p_178270_4_.func_181662_b(d0 + 1.0D, d1, d2 + 1.0D).func_181666_a(0.5F, 0.5F, 0.5F, 1.0F).func_187315_a((double)f36, (double)f38).func_187314_a(i2, j2).func_181675_d();
+                bufferBuilderIn.func_181662_b(d0, d1, d2 + 1.0D).func_181666_a(0.5F, 0.5F, 0.5F, 1.0F).func_187315_a((double)f35, (double)f38).func_187314_a(i2, j2).func_181675_d();
+                bufferBuilderIn.func_181662_b(d0, d1, d2).func_181666_a(0.5F, 0.5F, 0.5F, 1.0F).func_187315_a((double)f35, (double)f37).func_187314_a(i2, j2).func_181675_d();
+                bufferBuilderIn.func_181662_b(d0 + 1.0D, d1, d2).func_181666_a(0.5F, 0.5F, 0.5F, 1.0F).func_187315_a((double)f36, (double)f37).func_187314_a(i2, j2).func_181675_d();
+                bufferBuilderIn.func_181662_b(d0 + 1.0D, d1, d2 + 1.0D).func_181666_a(0.5F, 0.5F, 0.5F, 1.0F).func_187315_a((double)f36, (double)f38).func_187314_a(i2, j2).func_181675_d();
                 flag3 = true;
             }
 
@@ -179,14 +184,14 @@
                     ++j1;
                 }
 
-                BlockPos blockpos = p_178270_3_.func_177982_a(j1, 0, k1);
+                BlockPos blockpos = blockPosIn.func_177982_a(j1, 0, k1);
                 TextureAtlasSprite textureatlassprite1 = atextureatlassprite[1];
 
                 if (!flag)
                 {
-                    Block block = p_178270_1_.func_180495_p(blockpos).func_177230_c();
+                    IBlockState state = blockAccess.func_180495_p(blockpos);
 
-                    if (block == Blocks.field_150359_w || block == Blocks.field_150399_cn)
+                    if (state.func_193401_d(blockAccess, blockpos, EnumFacing.field_82609_l[i1+2].func_176734_d()) == net.minecraft.block.state.BlockFaceShape.SOLID)
                     {
                         textureatlassprite1 = this.field_187501_d;
                     }
@@ -244,24 +249,24 @@
                     float f28 = textureatlassprite1.func_94207_b((double)((1.0F - f39) * 16.0F * 0.5F));
                     float f29 = textureatlassprite1.func_94207_b((double)((1.0F - f40) * 16.0F * 0.5F));
                     float f30 = textureatlassprite1.func_94207_b(8.0D);
-                    int j = p_178270_2_.func_185889_a(p_178270_1_, blockpos);
+                    int j = blockStateIn.func_185889_a(blockAccess, blockpos);
                     int k = j >> 16 & 65535;
                     int l = j & 65535;
                     float f31 = i1 < 2 ? 0.8F : 0.6F;
                     float f32 = 1.0F * f31 * f;
                     float f33 = 1.0F * f31 * f1;
                     float f34 = 1.0F * f31 * f2;
-                    p_178270_4_.func_181662_b(d3, d1 + (double)f39, d4).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f41, (double)f28).func_187314_a(k, l).func_181675_d();
-                    p_178270_4_.func_181662_b(d5, d1 + (double)f40, d6).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f27, (double)f29).func_187314_a(k, l).func_181675_d();
-                    p_178270_4_.func_181662_b(d5, d1 + 0.0D, d6).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f27, (double)f30).func_187314_a(k, l).func_181675_d();
-                    p_178270_4_.func_181662_b(d3, d1 + 0.0D, d4).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f41, (double)f30).func_187314_a(k, l).func_181675_d();
+                    bufferBuilderIn.func_181662_b(d3, d1 + (double)f39, d4).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f41, (double)f28).func_187314_a(k, l).func_181675_d();
+                    bufferBuilderIn.func_181662_b(d5, d1 + (double)f40, d6).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f27, (double)f29).func_187314_a(k, l).func_181675_d();
+                    bufferBuilderIn.func_181662_b(d5, d1 + 0.0D, d6).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f27, (double)f30).func_187314_a(k, l).func_181675_d();
+                    bufferBuilderIn.func_181662_b(d3, d1 + 0.0D, d4).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f41, (double)f30).func_187314_a(k, l).func_181675_d();
 
                     if (textureatlassprite1 != this.field_187501_d)
                     {
-                        p_178270_4_.func_181662_b(d3, d1 + 0.0D, d4).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f41, (double)f30).func_187314_a(k, l).func_181675_d();
-                        p_178270_4_.func_181662_b(d5, d1 + 0.0D, d6).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f27, (double)f30).func_187314_a(k, l).func_181675_d();
-                        p_178270_4_.func_181662_b(d5, d1 + (double)f40, d6).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f27, (double)f29).func_187314_a(k, l).func_181675_d();
-                        p_178270_4_.func_181662_b(d3, d1 + (double)f39, d4).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f41, (double)f28).func_187314_a(k, l).func_181675_d();
+                        bufferBuilderIn.func_181662_b(d3, d1 + 0.0D, d4).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f41, (double)f30).func_187314_a(k, l).func_181675_d();
+                        bufferBuilderIn.func_181662_b(d5, d1 + 0.0D, d6).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f27, (double)f30).func_187314_a(k, l).func_181675_d();
+                        bufferBuilderIn.func_181662_b(d5, d1 + (double)f40, d6).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f27, (double)f29).func_187314_a(k, l).func_181675_d();
+                        bufferBuilderIn.func_181662_b(d3, d1 + (double)f39, d4).func_181666_a(f32, f33, f34, 1.0F).func_187315_a((double)f41, (double)f28).func_187314_a(k, l).func_181675_d();
                     }
                 }
             }
@@ -270,6 +275,301 @@
         }
     }
 
+    public boolean func_178270_a(IBlockAccess p_178270_1_, IBlockState p_178270_2_, BlockPos p_178270_3_, BufferBuilder p_178270_4_)
+    {
+        boolean flag3;
+
+        try
+        {
+            if (Config.isShaders())
+            {
+                SVertexBuilder.pushEntity(p_178270_2_, p_178270_3_, p_178270_1_, p_178270_4_);
+            }
+
+            BlockLiquid blockliquid = (BlockLiquid)p_178270_2_.func_177230_c();
+            boolean flag = p_178270_2_.func_185904_a() == Material.field_151587_i;
+            TextureAtlasSprite[] atextureatlassprite = flag ? this.field_178272_a : this.field_178271_b;
+            RenderEnv renderenv = p_178270_4_.getRenderEnv(p_178270_1_, p_178270_2_, p_178270_3_);
+            int i = CustomColors.getFluidColor(p_178270_1_, p_178270_2_, p_178270_3_, renderenv);
+            float f = (float)(i >> 16 & 255) / 255.0F;
+            float f1 = (float)(i >> 8 & 255) / 255.0F;
+            float f2 = (float)(i & 255) / 255.0F;
+            boolean flag1 = p_178270_2_.func_185894_c(p_178270_1_, p_178270_3_, EnumFacing.UP);
+            boolean flag2 = p_178270_2_.func_185894_c(p_178270_1_, p_178270_3_, EnumFacing.DOWN);
+            boolean[] aboolean = renderenv.getBorderFlags();
+            aboolean[0] = p_178270_2_.func_185894_c(p_178270_1_, p_178270_3_, EnumFacing.NORTH);
+            aboolean[1] = p_178270_2_.func_185894_c(p_178270_1_, p_178270_3_, EnumFacing.SOUTH);
+            aboolean[2] = p_178270_2_.func_185894_c(p_178270_1_, p_178270_3_, EnumFacing.WEST);
+            aboolean[3] = p_178270_2_.func_185894_c(p_178270_1_, p_178270_3_, EnumFacing.EAST);
+
+            if (flag1 || flag2 || aboolean[0] || aboolean[1] || aboolean[2] || aboolean[3])
+            {
+                flag3 = false;
+                float f3 = 0.5F;
+                float f4 = 1.0F;
+                float f5 = 0.8F;
+                float f6 = 0.6F;
+                Material material = p_178270_2_.func_185904_a();
+                float f7 = this.func_178269_a(p_178270_1_, p_178270_3_, material);
+                float f8 = this.func_178269_a(p_178270_1_, p_178270_3_.func_177968_d(), material);
+                float f9 = this.func_178269_a(p_178270_1_, p_178270_3_.func_177974_f().func_177968_d(), material);
+                float f10 = this.func_178269_a(p_178270_1_, p_178270_3_.func_177974_f(), material);
+                double d0 = (double)p_178270_3_.func_177958_n();
+                double d1 = (double)p_178270_3_.func_177956_o();
+                double d2 = (double)p_178270_3_.func_177952_p();
+                float f11 = 0.001F;
+
+                if (flag1)
+                {
+                    flag3 = true;
+                    float f12 = BlockLiquid.func_189544_a(p_178270_1_, p_178270_3_, material, p_178270_2_);
+                    TextureAtlasSprite textureatlassprite = f12 > -999.0F ? atextureatlassprite[1] : atextureatlassprite[0];
+                    p_178270_4_.setSprite(textureatlassprite);
+                    f7 -= 0.001F;
+                    f8 -= 0.001F;
+                    f9 -= 0.001F;
+                    f10 -= 0.001F;
+                    float f13;
+                    float f14;
+                    float f15;
+                    float f16;
+                    float f17;
+                    float f18;
+                    float f19;
+                    float f20;
+
+                    if (f12 < -999.0F)
+                    {
+                        f13 = textureatlassprite.func_94214_a(0.0D);
+                        f17 = textureatlassprite.func_94207_b(0.0D);
+                        f14 = f13;
+                        f18 = textureatlassprite.func_94207_b(16.0D);
+                        f15 = textureatlassprite.func_94214_a(16.0D);
+                        f19 = f18;
+                        f16 = f15;
+                        f20 = f17;
+                    }
+                    else
+                    {
+                        float f21 = MathHelper.func_76126_a(f12) * 0.25F;
+                        float f22 = MathHelper.func_76134_b(f12) * 0.25F;
+                        float f23 = 8.0F;
+                        f13 = textureatlassprite.func_94214_a((double)(8.0F + (-f22 - f21) * 16.0F));
+                        f17 = textureatlassprite.func_94207_b((double)(8.0F + (-f22 + f21) * 16.0F));
+                        f14 = textureatlassprite.func_94214_a((double)(8.0F + (-f22 + f21) * 16.0F));
+                        f18 = textureatlassprite.func_94207_b((double)(8.0F + (f22 + f21) * 16.0F));
+                        f15 = textureatlassprite.func_94214_a((double)(8.0F + (f22 + f21) * 16.0F));
+                        f19 = textureatlassprite.func_94207_b((double)(8.0F + (f22 - f21) * 16.0F));
+                        f16 = textureatlassprite.func_94214_a((double)(8.0F + (f22 - f21) * 16.0F));
+                        f20 = textureatlassprite.func_94207_b((double)(8.0F + (-f22 - f21) * 16.0F));
+                    }
+
+                    int k2 = p_178270_2_.func_185889_a(p_178270_1_, p_178270_3_);
+                    int l2 = k2 >> 16 & 65535;
+                    int i3 = k2 & 65535;
+                    float f24 = 1.0F * f;
+                    float f25 = 1.0F * f1;
+                    float f26 = 1.0F * f2;
+                    p_178270_4_.func_181662_b(d0 + 0.0D, d1 + (double)f7, d2 + 0.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f13, (double)f17).func_187314_a(l2, i3).func_181675_d();
+                    p_178270_4_.func_181662_b(d0 + 0.0D, d1 + (double)f8, d2 + 1.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f14, (double)f18).func_187314_a(l2, i3).func_181675_d();
+                    p_178270_4_.func_181662_b(d0 + 1.0D, d1 + (double)f9, d2 + 1.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f15, (double)f19).func_187314_a(l2, i3).func_181675_d();
+                    p_178270_4_.func_181662_b(d0 + 1.0D, d1 + (double)f10, d2 + 0.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f16, (double)f20).func_187314_a(l2, i3).func_181675_d();
+
+                    if (blockliquid.func_176364_g(p_178270_1_, p_178270_3_.func_177984_a()))
+                    {
+                        p_178270_4_.func_181662_b(d0 + 0.0D, d1 + (double)f7, d2 + 0.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f13, (double)f17).func_187314_a(l2, i3).func_181675_d();
+                        p_178270_4_.func_181662_b(d0 + 1.0D, d1 + (double)f10, d2 + 0.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f16, (double)f20).func_187314_a(l2, i3).func_181675_d();
+                        p_178270_4_.func_181662_b(d0 + 1.0D, d1 + (double)f9, d2 + 1.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f15, (double)f19).func_187314_a(l2, i3).func_181675_d();
+                        p_178270_4_.func_181662_b(d0 + 0.0D, d1 + (double)f8, d2 + 1.0D).func_181666_a(f24, f25, f26, 1.0F).func_187315_a((double)f14, (double)f18).func_187314_a(l2, i3).func_181675_d();
+                    }
+                }
+
+                if (flag2)
+                {
+                    float f38 = atextureatlassprite[0].func_94209_e();
+                    float f39 = atextureatlassprite[0].func_94212_f();
+                    float f40 = atextureatlassprite[0].func_94206_g();
+                    float f41 = atextureatlassprite[0].func_94210_h();
+                    int l1 = p_178270_2_.func_185889_a(p_178270_1_, p_178270_3_.func_177977_b());
+                    int i2 = l1 >> 16 & 65535;
+                    int j2 = l1 & 65535;
+                    p_178270_4_.func_181662_b(d0, d1, d2 + 1.0D).func_181666_a(f * 0.5F, f1 * 0.5F, f2 * 0.5F, 1.0F).func_187315_a((double)f38, (double)f41).func_187314_a(i2, j2).func_181675_d();
+                    p_178270_4_.func_181662_b(d0, d1, d2).func_181666_a(f * 0.5F, f1 * 0.5F, f2 * 0.5F, 1.0F).func_187315_a((double)f38, (double)f40).func_187314_a(i2, j2).func_181675_d();
+                    p_178270_4_.func_181662_b(d0 + 1.0D, d1, d2).func_181666_a(f * 0.5F, f1 * 0.5F, f2 * 0.5F, 1.0F).func_187315_a((double)f39, (double)f40).func_187314_a(i2, j2).func_181675_d();
+                    p_178270_4_.func_181662_b(d0 + 1.0D, d1, d2 + 1.0D).func_181666_a(f * 0.5F, f1 * 0.5F, f2 * 0.5F, 1.0F).func_187315_a((double)f39, (double)f41).func_187314_a(i2, j2).func_181675_d();
+                    flag3 = true;
+                }
+
+                for (int i1 = 0; i1 < 4; ++i1)
+                {
+                    int j1 = 0;
+                    int k1 = 0;
+
+                    if (i1 == 0)
+                    {
+                        --k1;
+                    }
+
+                    if (i1 == 1)
+                    {
+                        ++k1;
+                    }
+
+                    if (i1 == 2)
+                    {
+                        --j1;
+                    }
+
+                    if (i1 == 3)
+                    {
+                        ++j1;
+                    }
+
+                    BlockPos blockpos = p_178270_3_.func_177982_a(j1, 0, k1);
+                    TextureAtlasSprite textureatlassprite1 = atextureatlassprite[1];
+                    p_178270_4_.setSprite(textureatlassprite1);
+                    float f42 = 0.0F;
+                    float f43 = 0.0F;
+
+                    if (!flag)
+                    {
+                        IBlockState iblockstate = p_178270_1_.func_180495_p(blockpos);
+                        Block block = iblockstate.func_177230_c();
+
+                        if (block == Blocks.field_150359_w || block == Blocks.field_150399_cn || block == Blocks.field_150461_bJ || block == Blocks.field_180399_cE)
+                        {
+                            textureatlassprite1 = this.field_187501_d;
+                            p_178270_4_.setSprite(textureatlassprite1);
+                        }
+
+                        if (block == Blocks.field_150458_ak || block == Blocks.field_185774_da)
+                        {
+                            f42 = 0.9375F;
+                            f43 = 0.9375F;
+                        }
+
+                        if (block instanceof BlockSlab)
+                        {
+                            BlockSlab blockslab = (BlockSlab)block;
+
+                            if (!blockslab.func_176552_j() && iblockstate.func_177229_b(BlockSlab.field_176554_a) == BlockSlab.EnumBlockHalf.BOTTOM)
+                            {
+                                f42 = 0.5F;
+                                f43 = 0.5F;
+                            }
+                        }
+                    }
+
+                    if (aboolean[i1])
+                    {
+                        float f44;
+                        float f45;
+                        double d3;
+                        double d4;
+                        double d5;
+                        double d6;
+
+                        if (i1 == 0)
+                        {
+                            f44 = f7;
+                            f45 = f10;
+                            d3 = d0;
+                            d5 = d0 + 1.0D;
+                            d4 = d2 + 0.0010000000474974513D;
+                            d6 = d2 + 0.0010000000474974513D;
+                        }
+                        else if (i1 == 1)
+                        {
+                            f44 = f9;
+                            f45 = f8;
+                            d3 = d0 + 1.0D;
+                            d5 = d0;
+                            d4 = d2 + 1.0D - 0.0010000000474974513D;
+                            d6 = d2 + 1.0D - 0.0010000000474974513D;
+                        }
+                        else if (i1 == 2)
+                        {
+                            f44 = f8;
+                            f45 = f7;
+                            d3 = d0 + 0.0010000000474974513D;
+                            d5 = d0 + 0.0010000000474974513D;
+                            d4 = d2 + 1.0D;
+                            d6 = d2;
+                        }
+                        else
+                        {
+                            f44 = f10;
+                            f45 = f9;
+                            d3 = d0 + 1.0D - 0.0010000000474974513D;
+                            d5 = d0 + 1.0D - 0.0010000000474974513D;
+                            d4 = d2;
+                            d6 = d2 + 1.0D;
+                        }
+
+                        if (f44 > f42 || f45 > f43)
+                        {
+                            f42 = Math.min(f42, f44);
+                            f43 = Math.min(f43, f45);
+
+                            if (f42 > f11)
+                            {
+                                f42 -= f11;
+                            }
+
+                            if (f43 > f11)
+                            {
+                                f43 -= f11;
+                            }
+
+                            flag3 = true;
+                            float f27 = textureatlassprite1.func_94214_a(0.0D);
+                            float f28 = textureatlassprite1.func_94214_a(8.0D);
+                            float f29 = textureatlassprite1.func_94207_b((double)((1.0F - f44) * 16.0F * 0.5F));
+                            float f30 = textureatlassprite1.func_94207_b((double)((1.0F - f45) * 16.0F * 0.5F));
+                            float f31 = textureatlassprite1.func_94207_b(8.0D);
+                            float f32 = textureatlassprite1.func_94207_b((double)((1.0F - f42) * 16.0F * 0.5F));
+                            float f33 = textureatlassprite1.func_94207_b((double)((1.0F - f43) * 16.0F * 0.5F));
+                            int j = p_178270_2_.func_185889_a(p_178270_1_, blockpos);
+                            int k = j >> 16 & 65535;
+                            int l = j & 65535;
+                            float f34 = i1 < 2 ? 0.8F : 0.6F;
+                            float f35 = 1.0F * f34 * f;
+                            float f36 = 1.0F * f34 * f1;
+                            float f37 = 1.0F * f34 * f2;
+                            p_178270_4_.func_181662_b(d3, d1 + (double)f44, d4).func_181666_a(f35, f36, f37, 1.0F).func_187315_a((double)f27, (double)f29).func_187314_a(k, l).func_181675_d();
+                            p_178270_4_.func_181662_b(d5, d1 + (double)f45, d6).func_181666_a(f35, f36, f37, 1.0F).func_187315_a((double)f28, (double)f30).func_187314_a(k, l).func_181675_d();
+                            p_178270_4_.func_181662_b(d5, d1 + (double)f43, d6).func_181666_a(f35, f36, f37, 1.0F).func_187315_a((double)f28, (double)f33).func_187314_a(k, l).func_181675_d();
+                            p_178270_4_.func_181662_b(d3, d1 + (double)f42, d4).func_181666_a(f35, f36, f37, 1.0F).func_187315_a((double)f27, (double)f32).func_187314_a(k, l).func_181675_d();
+
+                            if (textureatlassprite1 != this.field_187501_d)
+                            {
+                                p_178270_4_.func_181662_b(d3, d1 + (double)f42, d4).func_181666_a(f35, f36, f37, 1.0F).func_187315_a((double)f27, (double)f32).func_187314_a(k, l).func_181675_d();
+                                p_178270_4_.func_181662_b(d5, d1 + (double)f43, d6).func_181666_a(f35, f36, f37, 1.0F).func_187315_a((double)f28, (double)f33).func_187314_a(k, l).func_181675_d();
+                                p_178270_4_.func_181662_b(d5, d1 + (double)f45, d6).func_181666_a(f35, f36, f37, 1.0F).func_187315_a((double)f28, (double)f30).func_187314_a(k, l).func_181675_d();
+                                p_178270_4_.func_181662_b(d3, d1 + (double)f44, d4).func_181666_a(f35, f36, f37, 1.0F).func_187315_a((double)f27, (double)f29).func_187314_a(k, l).func_181675_d();
+                            }
+                        }
+                    }
+                }
+
+                p_178270_4_.setSprite((TextureAtlasSprite)null);
+                boolean flag4 = flag3;
+                return flag4;
+            }
+
+            flag3 = false;
+        }
+        finally
+        {
+            if (Config.isShaders())
+            {
+                SVertexBuilder.popEntity(p_178270_4_);
+            }
+        }
+
+        return flag3;
+    }
+    
     private float func_178269_a(IBlockAccess p_178269_1_, BlockPos p_178269_2_, Material p_178269_3_)
     {
         int i = 0;
