--- ../src-base/minecraft/net/minecraft/client/renderer/RenderItem.java
+++ ../src-work/minecraft/net/minecraft/client/renderer/RenderItem.java
@@ -49,6 +49,7 @@
 import net.minecraft.item.ItemFishFood;
 import net.minecraft.item.ItemStack;
 import net.minecraft.tileentity.TileEntityStructure;
+import net.minecraft.util.BlockRenderLayer;
 import net.minecraft.util.EnumFacing;
 import net.minecraft.util.ReportedException;
 import net.minecraft.util.ResourceLocation;
@@ -57,6 +58,13 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import optifine.Config;
+import optifine.CustomColors;
+import optifine.CustomItems;
+import optifine.Reflector;
+import optifine.ReflectorForge;
+import shadersmod.client.Shaders;
+import shadersmod.client.ShadersRender;
 
 @SideOnly(Side.CLIENT)
 public class RenderItem implements IResourceManagerReloadListener
@@ -67,11 +75,19 @@
     private final ItemModelMesher field_175059_m;
     private final TextureManager field_175057_n;
     private final ItemColors field_184395_f;
+    
+    //op add
+    private ResourceLocation modelLocation = null;
+    private boolean renderItemGui = false;
+    public ModelManager modelManager = null;
 
     public RenderItem(TextureManager p_i46552_1_, ModelManager p_i46552_2_, ItemColors p_i46552_3_)
     {
         this.field_175057_n = p_i46552_1_;
-        this.field_175059_m = new ItemModelMesher(p_i46552_2_);
+        //op add
+        this.modelManager = p_i46552_2_;
+        //op end
+        this.field_175059_m = new net.minecraftforge.client.ItemModelMesherForge(p_i46552_2_);
         this.func_175041_b();
         this.field_184395_f = p_i46552_3_;
     }
@@ -106,15 +122,40 @@
         this.func_191967_a(p_191961_1_, -1, p_191961_2_);
     }
 
-    private void func_191965_a(IBakedModel p_191965_1_, int p_191965_2_)
+    public void func_191965_a(IBakedModel p_191965_1_, int p_191965_2_)
     {
         this.func_191967_a(p_191965_1_, p_191965_2_, ItemStack.field_190927_a);
     }
 
     private void func_191967_a(IBakedModel p_191967_1_, int p_191967_2_, ItemStack p_191967_3_)
     {
+       /* if (net.minecraftforge.common.ForgeModContainer.allowEmissiveItems)
+        {
+            net.minecraftforge.client.ForgeHooksClient.renderLitItem(this, model, color, stack);
+            return;
+        }
+        Tessellator tessellator = Tessellator.getInstance();
+        BufferBuilder bufferbuilder = tessellator.getBuffer();
+        bufferbuilder.begin(7, DefaultVertexFormats.ITEM);
+
+        for (EnumFacing enumfacing : EnumFacing.values())
+        {
+            this.renderQuads(bufferbuilder, model.getQuads((IBlockState)null, enumfacing, 0L), color, stack);
+        }
+
+        this.renderQuads(bufferbuilder, model.getQuads((IBlockState)null, (EnumFacing)null, 0L), color, stack);
+        tessellator.draw();*/
+
         Tessellator tessellator = Tessellator.func_178181_a();
         BufferBuilder bufferbuilder = tessellator.func_178180_c();
+        boolean flag = Minecraft.func_71410_x().func_147117_R().isTextureBound();
+        boolean flag1 = Config.isMultiTexture() && flag;
+
+        if (flag1)
+        {
+            bufferbuilder.setBlockLayer(BlockRenderLayer.SOLID);
+        }
+
         bufferbuilder.func_181668_a(7, DefaultVertexFormats.field_176599_b);
 
         for (EnumFacing enumfacing : EnumFacing.values())
@@ -124,6 +165,12 @@
 
         this.func_191970_a(bufferbuilder, p_191967_1_.func_188616_a((IBlockState)null, (EnumFacing)null, 0L), p_191967_2_, p_191967_3_);
         tessellator.func_78381_a();
+
+        if (flag1)
+        {
+            bufferbuilder.setBlockLayer((BlockRenderLayer)null);
+            GlStateManager.bindCurrentTexture();
+        }
     }
 
     public void func_180454_a(ItemStack p_180454_1_, IBakedModel p_180454_2_)
@@ -137,13 +184,22 @@
             {
                 GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
                 GlStateManager.func_179091_B();
+                //stack.getItem().getTileEntityItemStackRenderer().renderByItem(stack);
                 TileEntityItemStackRenderer.field_147719_a.func_179022_a(p_180454_1_);
             }
             else
             {
+                //op add
+                if (Config.isCustomItems())
+                {
+                    p_180454_2_ = CustomItems.getCustomItemModel(p_180454_1_, p_180454_2_, this.modelLocation, false);
+                    this.modelLocation = null;
+                }                
+                //op end
                 this.func_191961_a(p_180454_2_, p_180454_1_);
 
-                if (p_180454_1_.func_77962_s())
+                //if (stack.hasEffect())
+                if (p_180454_1_.func_77962_s() && (!Config.isCustomItems() || !CustomItems.renderCustomEffect(this, p_180454_1_, p_180454_2_))) // op add
                 {
                     this.func_191966_a(p_180454_2_);
                 }
@@ -153,7 +209,7 @@
         }
     }
 
-    private void func_191966_a(IBakedModel p_191966_1_)
+    private void renderEffectO(IBakedModel model)
     {
         GlStateManager.func_179132_a(false);
         GlStateManager.func_179143_c(514);
@@ -166,14 +222,14 @@
         float f = (float)(Minecraft.func_71386_F() % 3000L) / 3000.0F / 8.0F;
         GlStateManager.func_179109_b(f, 0.0F, 0.0F);
         GlStateManager.func_179114_b(-50.0F, 0.0F, 0.0F, 1.0F);
-        this.func_191965_a(p_191966_1_, -8372020);
+        this.func_191965_a(model, -8372020);
         GlStateManager.func_179121_F();
         GlStateManager.func_179094_E();
         GlStateManager.func_179152_a(8.0F, 8.0F, 8.0F);
         float f1 = (float)(Minecraft.func_71386_F() % 4873L) / 4873.0F / 8.0F;
         GlStateManager.func_179109_b(-f1, 0.0F, 0.0F);
         GlStateManager.func_179114_b(10.0F, 0.0F, 0.0F, 1.0F);
-        this.func_191965_a(p_191966_1_, -8372020);
+        this.func_191965_a(model, -8372020);
         GlStateManager.func_179121_F();
         GlStateManager.func_179128_n(5888);
         GlStateManager.func_187401_a(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
@@ -181,21 +237,116 @@
         GlStateManager.func_179143_c(515);
         GlStateManager.func_179132_a(true);
         this.field_175057_n.func_110577_a(TextureMap.field_110575_b);
-    }
+    }  
+    private void func_191966_a(IBakedModel p_191966_1_)
+    {
+        if (!Config.isCustomItems() || CustomItems.isUseGlint())
+        {
+            if (!Config.isShaders() || !Shaders.isShadowPass)
+            {
+                GlStateManager.func_179132_a(false);
+                GlStateManager.func_179143_c(514);
+                GlStateManager.func_179140_f();
+                GlStateManager.func_187401_a(GlStateManager.SourceFactor.SRC_COLOR, GlStateManager.DestFactor.ONE);
+                this.field_175057_n.func_110577_a(field_110798_h);
 
+                if (Config.isShaders() && !this.renderItemGui)
+                {
+                    ShadersRender.renderEnchantedGlintBegin();
+                }
+
+                GlStateManager.func_179128_n(5890);
+                GlStateManager.func_179094_E();
+                GlStateManager.func_179152_a(8.0F, 8.0F, 8.0F);
+                float f = (float)(Minecraft.func_71386_F() % 3000L) / 3000.0F / 8.0F;
+                GlStateManager.func_179109_b(f, 0.0F, 0.0F);
+                GlStateManager.func_179114_b(-50.0F, 0.0F, 0.0F, 1.0F);
+                this.func_191965_a(p_191966_1_, -8372020);
+                GlStateManager.func_179121_F();
+                GlStateManager.func_179094_E();
+                GlStateManager.func_179152_a(8.0F, 8.0F, 8.0F);
+                float f1 = (float)(Minecraft.func_71386_F() % 4873L) / 4873.0F / 8.0F;
+                GlStateManager.func_179109_b(-f1, 0.0F, 0.0F);
+                GlStateManager.func_179114_b(10.0F, 0.0F, 0.0F, 1.0F);
+                this.func_191965_a(p_191966_1_, -8372020);
+                GlStateManager.func_179121_F();
+                GlStateManager.func_179128_n(5888);
+                GlStateManager.func_187401_a(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
+                GlStateManager.func_179145_e();
+                GlStateManager.func_179143_c(515);
+                GlStateManager.func_179132_a(true);
+                this.field_175057_n.func_110577_a(TextureMap.field_110575_b);
+
+                if (Config.isShaders() && !this.renderItemGui)
+                {
+                    ShadersRender.renderEnchantedGlintEnd();
+                }
+            }
+        }
+    }
+    
     private void func_175038_a(BufferBuilder p_175038_1_, BakedQuad p_175038_2_)
     {
         Vec3i vec3i = p_175038_2_.func_178210_d().func_176730_m();
         p_175038_1_.func_178975_e((float)vec3i.func_177958_n(), (float)vec3i.func_177956_o(), (float)vec3i.func_177952_p());
     }
 
+    private void renderQuad0(BufferBuilder renderer, BakedQuad quad, int color)
+    {
+        renderer.func_178981_a(quad.func_178209_a());
+        renderer.func_178968_d(color);
+        this.func_175038_a(renderer, quad);
+    }
+
     private void func_191969_a(BufferBuilder p_191969_1_, BakedQuad p_191969_2_, int p_191969_3_)
     {
-        p_191969_1_.func_178981_a(p_191969_2_.func_178209_a());
-        p_191969_1_.func_178968_d(p_191969_3_);
-        this.func_175038_a(p_191969_1_, p_191969_2_);
+        
+            if (p_191969_1_.isMultiTexture())
+            {
+                p_191969_1_.func_178981_a(p_191969_2_.getVertexDataSingle());
+                p_191969_1_.putSprite(p_191969_2_.func_187508_a());
+            }
+            else
+            {
+                p_191969_1_.func_178981_a(p_191969_2_.func_178209_a());
+            }
+
+            if (Reflector.ForgeHooksClient_putQuadColor.exists())
+            {
+                Reflector.call(Reflector.ForgeHooksClient_putQuadColor, p_191969_1_, p_191969_2_, p_191969_3_);
+            }
+            else
+            {
+                p_191969_1_.func_178968_d(p_191969_3_);
+            }
+
+            this.func_175038_a(p_191969_1_, p_191969_2_);
     }
 
+    private void func_191969_a(BufferBuilder p_191969_1_, BakedQuad p_191969_2_, int p_191969_3_)
+    {
+        if (p_191969_1_.isMultiTexture())
+        {
+            p_191969_1_.func_178981_a(p_191969_2_.getVertexDataSingle());
+            p_191969_1_.putSprite(p_191969_2_.func_187508_a());
+        }
+        else
+        {
+            p_191969_1_.func_178981_a(p_191969_2_.func_178209_a());
+        }
+
+        if (Reflector.ForgeHooksClient_putQuadColor.exists())
+        {
+            Reflector.call(Reflector.ForgeHooksClient_putQuadColor, p_191969_1_, p_191969_2_, p_191969_3_);
+        }
+        else
+        {
+            p_191969_1_.func_178968_d(p_191969_3_);
+        }
+
+        this.func_175038_a(p_191969_1_, p_191969_2_);
+    }
+    
     public void func_191970_a(BufferBuilder p_191970_1_, List<BakedQuad> p_191970_2_, int p_191970_3_, ItemStack p_191970_4_)
     {
         boolean flag = p_191970_3_ == -1 && !p_191970_4_.func_190926_b();
@@ -210,6 +361,12 @@
             {
                 k = this.field_184395_f.func_186728_a(p_191970_4_, bakedquad.func_178211_c());
 
+                if (Config.isCustomColors())
+                {
+                    k = CustomColors.getColorFromItemStack(p_191970_4_, bakedquad.func_178211_c(), k);
+                }
+
+                
                 if (EntityRenderer.field_78517_a)
                 {
                     k = TextureUtil.func_177054_c(k);
@@ -217,8 +374,20 @@
 
                 k = k | -16777216;
             }
-
-            this.func_191969_a(p_191970_1_, bakedquad, k);
+          //op add
+            if (p_191970_1_.isMultiTexture()) {
+                p_191970_1_.func_178981_a(bakedquad.getVertexDataSingle());
+                p_191970_1_.putSprite(bakedquad.func_187508_a());
+            } else {
+                p_191970_1_.func_178981_a(bakedquad.func_178209_a());
+            }
+          
+          //op end
+            //net.minecraftforge.client.model.pipeline.LightUtil.renderQuadColor(renderer, bakedquad, k);
+            net.minecraftforge.client.ForgeHooksClient.putQuadColor(p_191970_1_, bakedquad, k);
+            this.func_175038_a(p_191970_1_,bakedquad);
+            
+            //this.func_191969_a(renderer, bakedquad, k);
         }
     }
 
@@ -239,11 +408,32 @@
 
     public IBakedModel func_184393_a(ItemStack p_184393_1_, @Nullable World p_184393_2_, @Nullable EntityLivingBase p_184393_3_)
     {
+        //IBakedModel ibakedmodel = this.itemModelMesher.getItemModel(stack);
+        //return ibakedmodel.getOverrides().handleItemState(ibakedmodel, stack, worldIn, entitylivingbaseIn);
         IBakedModel ibakedmodel = this.field_175059_m.func_178089_a(p_184393_1_);
         Item item = p_184393_1_.func_77973_b();
 
-        if (item != null && item.func_185040_i())
+        if (Config.isCustomItems())
         {
+            if (item != null && item.func_185040_i())
+            {
+                this.modelLocation = ibakedmodel.func_188617_f().func_188021_a(p_184393_1_, p_184393_2_, p_184393_3_);
+            }
+
+            IBakedModel ibakedmodel1 = CustomItems.getCustomItemModel(p_184393_1_, ibakedmodel, this.modelLocation, true);
+
+            if (ibakedmodel1 != ibakedmodel)
+            {
+                return ibakedmodel1;
+            }
+        }
+
+        if (Reflector.ForgeItemOverrideList_handleItemState.exists())
+        {
+            return (IBakedModel)Reflector.call(ibakedmodel.func_188617_f(), Reflector.ForgeItemOverrideList_handleItemState, ibakedmodel, p_184393_1_, p_184393_2_, p_184393_3_);
+        }
+        else if (item != null && item.func_185040_i())
+        {
             ResourceLocation resourcelocation = ibakedmodel.func_188617_f().func_188021_a(p_184393_1_, p_184393_2_, p_184393_3_);
             return resourcelocation == null ? ibakedmodel : this.field_175059_m.func_178083_a().func_174953_a(new ModelResourceLocation(resourcelocation, "inventory"));
         }
@@ -274,15 +464,12 @@
             GlStateManager.func_179147_l();
             GlStateManager.func_187428_a(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);
             GlStateManager.func_179094_E();
-            ItemCameraTransforms itemcameratransforms = p_184394_2_.func_177552_f();
-            ItemCameraTransforms.func_188034_a(itemcameratransforms.func_181688_b(p_184394_3_), p_184394_4_);
+            // TODO: check if negative scale is a thing
+            p_184394_2_ = net.minecraftforge.client.ForgeHooksClient.handleCameraTransforms(p_184394_2_, p_184394_3_, p_184394_4_);
 
-            if (this.func_183005_a(itemcameratransforms.func_181688_b(p_184394_3_)))
-            {
-                GlStateManager.func_187407_a(GlStateManager.CullFace.FRONT);
-            }
-
+            CustomItems.setRenderOffHand(p_184394_4_); // op add
             this.func_180454_a(p_184394_1_, p_184394_2_);
+            CustomItems.setRenderOffHand(false); // op add
             GlStateManager.func_187407_a(GlStateManager.CullFace.BACK);
             GlStateManager.func_179121_F();
             GlStateManager.func_179101_C();
@@ -304,6 +491,9 @@
 
     protected void func_191962_a(ItemStack p_191962_1_, int p_191962_2_, int p_191962_3_, IBakedModel p_191962_4_)
     {
+        //op add
+        this.renderItemGui = true;
+        //op end
         GlStateManager.func_179094_E();
         this.field_175057_n.func_110577_a(TextureMap.field_110575_b);
         this.field_175057_n.func_110581_b(TextureMap.field_110575_b).func_174936_b(false, false);
@@ -314,7 +504,8 @@
         GlStateManager.func_187401_a(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA);
         GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
         this.func_180452_a(p_191962_2_, p_191962_3_, p_191962_4_.func_177556_c());
-        p_191962_4_.func_177552_f().func_181689_a(ItemCameraTransforms.TransformType.GUI);
+        p_191962_4_ = net.minecraftforge.client.ForgeHooksClient.handleCameraTransforms(p_191962_4_, ItemCameraTransforms.TransformType.GUI, false);
+        
         this.func_180454_a(p_191962_1_, p_191962_4_);
         GlStateManager.func_179118_c();
         GlStateManager.func_179101_C();
@@ -322,6 +513,9 @@
         GlStateManager.func_179121_F();
         this.field_175057_n.func_110577_a(TextureMap.field_110575_b);
         this.field_175057_n.func_110581_b(TextureMap.field_110575_b).func_174935_a();
+        //op add
+        this.renderItemGui = false;        
+        //op end
     }
 
     private void func_180452_a(int p_180452_1_, int p_180452_2_, boolean p_180452_3_)
@@ -367,6 +561,7 @@
                         return String.valueOf((Object)p_184391_2_.func_77973_b());
                     }
                 });
+                crashreportcategory.func_189529_a("Registry Name", () -> String.valueOf(p_184391_2_.func_77973_b().getRegistryName()));
                 crashreportcategory.func_189529_a("Item Aux", new ICrashReportDetail<String>()
                 {
                     public String call() throws Exception
@@ -413,9 +608,13 @@
                 p_180453_1_.func_175063_a(s, (float)(p_180453_3_ + 19 - 2 - p_180453_1_.func_78256_a(s)), (float)(p_180453_4_ + 6 + 3), 16777215);
                 GlStateManager.func_179145_e();
                 GlStateManager.func_179126_j();
+                // Fixes opaque cooldown overlay a bit lower
+                // TODO: check if enabled blending still screws things up down the line.
+                GlStateManager.func_179147_l();
             }
 
-            if (p_180453_2_.func_77951_h())
+            //if (stack.getItem().showDurabilityBar(stack))
+            if(ReflectorForge.isItemDamaged(p_180453_2_))
             {
                 GlStateManager.func_179140_f();
                 GlStateManager.func_179097_i();
@@ -424,11 +623,46 @@
                 GlStateManager.func_179084_k();
                 Tessellator tessellator = Tessellator.func_178181_a();
                 BufferBuilder bufferbuilder = tessellator.func_178180_c();
+               /*  //op remove
+                double health = stack.getItem().getDurabilityForDisplay(stack);
+                int rgbfordisplay = stack.getItem().getRGBDurabilityForDisplay(stack);
+                int i = Math.round(13.0F - (float)health * 13.0F);
+                int j = rgbfordisplay;
+                */
+                
+                //op add
                 float f = (float)p_180453_2_.func_77952_i();
                 float f1 = (float)p_180453_2_.func_77958_k();
                 float f2 = Math.max(0.0F, (f1 - f) / f1);
                 int i = Math.round(13.0F - f * 13.0F / f1);
                 int j = MathHelper.func_181758_c(f2 / 3.0F, 1.0F, 1.0F);
+
+                if (Reflector.ForgeItem_getDurabilityForDisplay.exists() && Reflector.ForgeItem_getRGBDurabilityForDisplay.exists())
+                {
+                    double d0 = Reflector.callDouble(p_180453_2_.func_77973_b(), Reflector.ForgeItem_getDurabilityForDisplay, p_180453_2_);
+                    int k = Reflector.callInt(p_180453_2_.func_77973_b(), Reflector.ForgeItem_getRGBDurabilityForDisplay, p_180453_2_);
+                    i = Math.round(13.0F - (float)d0 * 13.0F);
+                    j = k;
+                }
+
+                if (Config.isCustomColors())
+                {
+                    j = CustomColors.getDurabilityColor(f2, j);
+                }
+
+                if (Reflector.ForgeItem_getDurabilityForDisplay.exists() && Reflector.ForgeItem_getRGBDurabilityForDisplay.exists())
+                {
+                    double d1 = Reflector.callDouble(p_180453_2_.func_77973_b(), Reflector.ForgeItem_getDurabilityForDisplay, p_180453_2_);
+                    int l = Reflector.callInt(p_180453_2_.func_77973_b(), Reflector.ForgeItem_getRGBDurabilityForDisplay, p_180453_2_);
+                    i = Math.round(13.0F - (float)d1 * 13.0F);
+                    j = l;
+                }
+
+                if (Config.isCustomColors())
+                {
+                    j = CustomColors.getDurabilityColor(f2, j);
+                }                
+                //op end
                 this.func_181565_a(bufferbuilder, p_180453_3_ + 2, p_180453_4_ + 13, 13, 2, 0, 0, 0, 255);
                 this.func_181565_a(bufferbuilder, p_180453_3_ + 2, p_180453_4_ + 13, i, 1, j >> 16 & 255, j >> 8 & 255, j & 255, 255);
                 GlStateManager.func_179147_l();
@@ -1099,6 +1333,7 @@
         this.func_175029_a(Blocks.field_185779_df, TileEntityStructure.Mode.LOAD.func_185110_a(), "structure_block");
         this.func_175029_a(Blocks.field_185779_df, TileEntityStructure.Mode.CORNER.func_185110_a(), "structure_block");
         this.func_175029_a(Blocks.field_185779_df, TileEntityStructure.Mode.DATA.func_185110_a(), "structure_block");
+        net.minecraftforge.client.model.ModelLoader.onRegisterItems(this.field_175059_m);
     }
 
     public void func_110549_a(IResourceManager p_110549_1_)
